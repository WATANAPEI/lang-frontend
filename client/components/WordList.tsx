import React from "react";
// @ts-ignore
import useWordsApi from "../hooks/useWordsApi.tsx";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import Paper from "@material-ui/core/Paper";
import CircularProgress from "@material-ui/core/CircularProgress";
import Container from "@material-ui/core/Container";

interface WordResponse {
  id: number;
  word: string;
  meaning: string;
  wordLanguageID: number;
  meaningLanguageID: number;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    wordList: {
      color: "red",
      listStyle: "none"
    },
    wordListContainer: {
      display: "flex",
      flexWrap: "wrap",
      overflow: "auto",
      height: "100%"
    },
    paper: {
      padding: theme.spacing(2),
      margin: "auto"
    },
    loadingMessage: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center"
    }
  })
);

function WordList() {
  let backendUrl: string;
  console.log(`NODE_ENV: ${process.env.NODE_ENV}`);
  if (process.env.NODE_ENV === "production") {
    backendUrl = "https://wpei.dev/lang/api/v1/words";
  } else {
    backendUrl = "http://127.0.0.1:3000/words";
  }
  const [{ words, isLoading, isError }, doFetch] = useWordsApi(
    backendUrl,
    [{
      id: -1,
      word: "initialize_error",
      meaning: "initialize_error",
      wordLanguageID: -1,
      meaningLanguageID: -1
    }
  ]);
  const classes = useStyles();
  const wordLists = words.map((word: WordResponse) =>
    <Paper key={word.id.toString()} className={classes.paper}>
      <Grid container direction="row" spacing={2}>
        <Grid item>
          <h1>{word.id}</h1>
        </Grid>
        <Grid item xs={12} sm container>
          <Grid item xs container direction="column" spacing={2}>
            <Grid item xs>
              <h3>{word.word}</h3>
              <h3>{word.meaning}</h3>
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </Paper>
  );
  return (
    <div className={classes.wordListContainer}>
      {isError && (
        <Container id="loadingErrorMessage" className={classes.loadingMessage}>
          Something went wrong...
        </Container>
      )}
      {!isError &&
        (isLoading ? (
          <Container id="loadingMessage" className={classes.loadingMessage}>
            <CircularProgress />
          </Container>
        ) : (
          wordLists
        ))}
    </div>
  );
}

export default WordList;
