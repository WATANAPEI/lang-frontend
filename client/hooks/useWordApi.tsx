import React, { useState, useEffect } from "react";

export interface WordResponse {
  id: number;
  word: string;
  meaning: string;
  wordLanguageID: number;
  meaningLanguageID: number;
}

export interface ReturnData {
  word: WordResponse;
  isLoading: boolean;
  isError: boolean;
}

interface RawData {
  id: number;
  word: string;
  meaning: string;
  word_lang_id: number;
  meaning_lang_id: number;
  created_by: string;
  last_updated_by: string;
  created_at: string;
  updated_at: string;
}

interface ReturnJson {
  status: string;
  message: string;
  data: RawData;
}

interface UseWordApi {
  (initialUrl: string, initialWord: WordResponse): [
    ReturnData,
    React.Dispatch<React.SetStateAction<string>>
  ];
}

const useWordApi: UseWordApi = (initialUrl, initialWord) => {
  const [word, setWord] = useState(initialWord);
  const [url, setUrl] = useState(initialUrl);
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    const fetchData = async () => {
      setIsError(false);
      setIsLoading(true);

      try {
        const response = await fetch(url);
        console.log(`reponse: ${JSON.stringify(response)}`);
        const json: ReturnJson = await response.json();
        console.log(`json: ${JSON.stringify(json)}`);
        const data: RawData = json.data;
        console.log(`data: ${JSON.stringify(data)}`);
        const word: WordResponse =
        {
          id: data.id,
          word: data.word,
          meaning: data.meaning,
          wordLanguageID: data.word_lang_id,
          meaningLanguageID: data.meaning_lang_id
        };
//      console.log(`json: ${JSON.stringify(word)}`);
//      console.log(`data: ${data}`);
//      console.log(`word: ${word}`);
        setWord(word);
      } catch (error) {
        setIsError(true);
        console.log(`error: ${JSON.stringify(error)}`);
      }
      setIsLoading(false);
    };

    fetchData();
  }, [url]);

  return [{ word, isLoading, isError }, setUrl];
};

export default useWordApi;
