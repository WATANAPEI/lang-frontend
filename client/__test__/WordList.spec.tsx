import React from "react";
import { act } from "react-dom/test-utils";
// @ts-ignore
import WordList from "../components/WordList.tsx";
import { createMount } from "@material-ui/core/test-utils";
// @ts-ignore
import { WordResponse } from "../hooks/useWordApi.tsx";
// @ts-ignore
import { mockWordsFactory, MockFetch } from "./MockFactory.tsx";
import "isomorphic-fetch";
import Paper from "@material-ui/core/Paper";

describe("<WordList />", () => {
  it("has Progress indication and Paper will appear after loading", done => {
    const mockUrlList = [
        "http://127.0.0.1:3000/words"
    ];
    let mockWordsArray: WordResponse[][] = [];
    let mockFetchArray: MockFetch[] = [];
    for (let i = 0; i < mockUrlList.length; i++) {
      [mockWordsArray[i], mockFetchArray[i]] = mockWordsFactory(
        "success",
        i,
        mockUrlList[i]
      );
    }
    window.fetch = jest.fn().
      mockImplementationOnce(mockFetchArray[0]);
    const fetchSpy = jest.spyOn(window, "fetch");

    act(() => {
      const wrapper = createMount()(<WordList />);
      setImmediate(() => {
        //console.log(wrapper.debug());
        expect(wrapper.find("#loadingMessage").exists()).toEqual(true);
        expect(wrapper).toMatchSnapshot();
        wrapper.update();
        //console.log(wrapper.debug());
        //console.log(mockWordsArray[0]);
        expect(wrapper.find(Paper)).toHaveLength(mockWordsArray[0].length);
        expect(wrapper.find(".word_id").exists()).toEqual(true);
        expect(wrapper.find(".word_word").exists()).toEqual(true);
        expect(wrapper.find(".word_meaning").exists()).toEqual(true);
        expect(fetchSpy).toHaveBeenCalledTimes(1);
        expect(fetchSpy).toHaveBeenLastCalledWith(mockUrlList[0]);
        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
        done();
      });
    });
  });
  it("render error message when data loading failed", done => {
    const mockUrlList = [
        "http://127.0.0.1:3000/words"
    ];
    let mockWordsArray: WordResponse[][] = [];
    let mockFetchArray: MockFetch[] = [];
    for (let i = 0; i < mockUrlList.length; i++) {
      [mockWordsArray[i], mockFetchArray[i]] = mockWordsFactory(
        "failed",
        i,
        mockUrlList[i]
      );
    }
    window.fetch = jest.fn().
      mockImplementationOnce(mockFetchArray[0]);
    const fetchSpy = jest.spyOn(window, "fetch");

    act(() => {
      const wrapper = createMount()(<WordList />);
      setImmediate(() => {
        wrapper.update();
        //console.log(wrapper.debug());
        expect(wrapper.find("#loadingErrorMessage").exists()).toEqual(true);
        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
        done();
      });
    });
  });
});
