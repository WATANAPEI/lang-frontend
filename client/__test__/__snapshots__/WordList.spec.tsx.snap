// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WordList /> has Progress indication and Paper will appear after loading 1`] = `
<WordList>
  <div
    className="makeStyles-wordListContainer-2"
  >
    <WithStyles(ForwardRef(Container))
      className="makeStyles-loadingMessage-4"
      id="loadingMessage"
    >
      <ForwardRef(Container)
        className="makeStyles-loadingMessage-4"
        classes={
          Object {
            "fixed": "MuiContainer-fixed",
            "maxWidthLg": "MuiContainer-maxWidthLg",
            "maxWidthMd": "MuiContainer-maxWidthMd",
            "maxWidthSm": "MuiContainer-maxWidthSm",
            "maxWidthXl": "MuiContainer-maxWidthXl",
            "maxWidthXs": "MuiContainer-maxWidthXs",
            "root": "MuiContainer-root",
          }
        }
        id="loadingMessage"
      >
        <div
          className="MuiContainer-root makeStyles-loadingMessage-4 MuiContainer-maxWidthLg"
          id="loadingMessage"
        >
          <WithStyles(ForwardRef(CircularProgress))>
            <ForwardRef(CircularProgress)
              classes={
                Object {
                  "circle": "MuiCircularProgress-circle",
                  "circleDisableShrink": "MuiCircularProgress-circleDisableShrink",
                  "circleIndeterminate": "MuiCircularProgress-circleIndeterminate",
                  "circleStatic": "MuiCircularProgress-circleStatic",
                  "colorPrimary": "MuiCircularProgress-colorPrimary",
                  "colorSecondary": "MuiCircularProgress-colorSecondary",
                  "indeterminate": "MuiCircularProgress-indeterminate",
                  "root": "MuiCircularProgress-root",
                  "static": "MuiCircularProgress-static",
                  "svg": "MuiCircularProgress-svg",
                }
              }
            >
              <div
                className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                role="progressbar"
                style={
                  Object {
                    "height": 40,
                    "width": 40,
                  }
                }
              >
                <svg
                  className="MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                    cx={44}
                    cy={44}
                    fill="none"
                    r={20.2}
                    strokeWidth={3.6}
                    style={Object {}}
                  />
                </svg>
              </div>
            </ForwardRef(CircularProgress)>
          </WithStyles(ForwardRef(CircularProgress))>
        </div>
      </ForwardRef(Container)>
    </WithStyles(ForwardRef(Container))>
  </div>
</WordList>
`;
